### PR TITLE
Integrate keepers 2.1 contracts

### PR DESCRIPTION
**Pull request by amirylm**
This PR contains WIP integration with 2.1 contract.

In addition, a stable reference of 2.0 contract was copied from `develop`.

### CODE REVIEW

File: core/services/ocr2/plugins/ocr2keeper/evm/encoder.go [hunk 1]
```diff
@@ -26,6 +26,7 @@ var (
 	ErrUnexpectedResult = fmt.Errorf("unexpected result struct")
 )
 
+// TODO
 type EVMAutomationUpkeepResult20 struct {
```
Comment (FelixFan1992) [inline core/services/ocr2/plugins/ocr2keeper/evm/encoder.go:30]: can we rename this too?

File: core/services/ocr2/plugins/ocr2keeper/util.go [hunk 2]
```diff
@@ -72,7 +72,7 @@ func EVMDependencies(spec job.Job, db *sqlx.DB, lggr logger.Logger, set evm.Chai
 	}
 
 	rAddr := ethkey.MustEIP55Address(oSpec.ContractID).Address()
-	if registry, err = kevm.NewEVMRegistryServiceV2_0(rAddr, chain, mc, lggr); err != nil {
+	if registry, err = kevm.NewEVMRegistryServiceV2_1(rAddr, chain, mc, lggr); err != nil {
```
Comment (infiloop2) [inline core/services/ocr2/plugins/ocr2keeper/util.go:75]: How is the clash between this file and `core/services/ocr2/plugins/ocr2keeper20/util.go` avoided? I see they are in the same package and have same function names. 

These functions are called in the top level delegate https://github.com/smartcontractkit/chainlink/blob/develop/core/services/ocr2/delegate.go#L649-L659 

would be good to plug this 2.1 change within delegate too within this PR to clarify how we choose 2.0 vs 2.1

File: core/services/ocr2/plugins/ocr2keeper/evm/abi.go [hunk 3]
```diff
@@ -27,13 +24,45 @@ const (
...
-func (rp *evmRegistryPackerV2_0) UnpackCheckResult(key ocr2keepers.UpkeepKey, raw string) (EVMAutomationUpkeepResult20, error) {
+type evmRegistryPackerV2_1 struct {
+	abi abi.ABI
+}
+
+func NewEvmRegistryPackerV2_1(abi abi.ABI) *evmRegistryPackerV2_1 {
+	return &evmRegistryPackerV2_1{abi: abi}
+}
+
+func (rp *evmRegistryPackerV2_1) UnpackCheckResult(key ocr2keepers.UpkeepKey, raw string) (EVMAutomationUpkeepResult21, error) {
```
Comment (RyanRHall) [inline core/services/ocr2/plugins/ocr2keeper/evm/abi.go:63]: this is unnecessary with registry version 2.1

File: core/scripts/chaincli/handler/report.go [hunk 4]
```diff
@@ -20,7 +20,7 @@ import (
 	ocr2keepers "github.com/smartcontractkit/ocr2keepers/pkg"
 
 	"github.com/smartcontractkit/chainlink/v2/core/gethwrappers/generated/keeper_registry_wrapper2_0"
-	"github.com/smartcontractkit/chainlink/v2/core/services/ocr2/plugins/ocr2keeper/evm"
+	evm "github.com/smartcontractkit/chainlink/v2/core/services/ocr2/plugins/ocr2keeper/evm20"
```
Comment (infiloop2) [inline core/scripts/chaincli/handler/report.go:23]: Not sure if it will be easier to just replace all instances of 'evm' with 'evm20'? Would make it clearer in code that it only supports 2.0 version of evm

File: core/services/ocr2/plugins/ocr2keeper/util.go [hunk 5]
```diff
@@ -17,7 +17,7 @@ import (
 	"github.com/smartcontractkit/chainlink/v2/core/services/job"
 	"github.com/smartcontractkit/chainlink/v2/core/services/keystore/keys/ethkey"
 	"github.com/smartcontractkit/chainlink/v2/core/services/ocr2/models"
-	kevm "github.com/smartcontractkit/chainlink/v2/core/services/ocr2/plugins/ocr2keeper/evm"
+	kevm "github.com/smartcontractkit/chainlink/v2/core/services/ocr2/plugins/ocr2keeper/evm20"
```
Comment (infiloop2) [inline core/services/ocr2/plugins/ocr2keeper/util.go:20]: this needs some more thought but since these utils are common across both versions, version can be sent as a parameter to EVMDependencies which chooses which dependencies to return. Will be more important when this is connected to the delegate (are you planning to do that in a separate PR?)

File: core/services/ocr2/plugins/ocr2keeper/evm/abi.go [hunk 6]
```diff
@@ -192,30 +168,18 @@ func (rp *evmRegistryPackerV2_0) UnpackTransmitTxInput(raw []byte) ([]ocr2keeper
...
+	var cfg iregistry21.KeeperRegistryBase21LogTriggerConfig
 
-type performDataWrapper struct {
-	Result performDataStruct
-}
-type performDataStruct struct {
-	CheckBlockNumber uint32   `abi:"checkBlockNumber"`
-	CheckBlockhash   [32]byte `abi:"checkBlockhash"`
-	PerformData      []byte   `abi:"performData"`
+	out, err := rp.abi.Methods["getLogTriggerConfig"].Outputs.UnpackValues(raw)
```
Comment (infiloop2) [inline core/services/ocr2/plugins/ocr2keeper/evm/abi.go:175]: this will also likely change but can be fixed later.

File: core/services/ocr2/plugins/ocr2keeper/evm/encoder.go [hunk 7]
```diff
@@ -112,32 +117,32 @@ func (enc EVMAutomationEncoder20) DecodeReport(report []byte) ([]ocr2keepers.Upk
...
+		wei  *big.Int
+		link *big.Int
 	)
 
 	if upkeepIds, ok = m[mKeys[2]].([]*big.Int); !ok {
 		return res, fmt.Errorf("upkeep ids of incorrect type in report")
 	}
 
-	// TODO: a type assertion on `wrappedPerform` did not work, even with the
+	// TODO: a type assertion on `wrappedTrigger` did not work, even with the
```
Comment (infiloop2) [inline core/services/ocr2/plugins/ocr2keeper/evm/encoder.go:130]: lol. same problem different struct

File: core/services/ocr2/plugins/ocr2keeper/evm20/log_provider.go [hunk 8]
```diff
@@ -442,7 +440,7 @@ func (c *LogProvider) getCheckBlockNumberFromTxHash(txHash common.Hash, id ocr2k
 
 	for _, upkeep := range decodedReport {
 		// TODO: the log provider should be in the evm package for isolation
```
Comment (infiloop2) [inline core/services/ocr2/plugins/ocr2keeper/evm20/log_provider.go:442]: this got automatically completed lol

File: core/services/ocr2/plugins/ocr2keeper/evm20/log_provider.go [hunk 9]
```diff
@@ -18,8 +18,6 @@ import (
 	"github.com/smartcontractkit/chainlink/v2/core/chains/evm/logpoller"
 	registry "github.com/smartcontractkit/chainlink/v2/core/gethwrappers/generated/keeper_registry_wrapper2_0"
 	"github.com/smartcontractkit/chainlink/v2/core/logger"
-	"github.com/smartcontractkit/chainlink/v2/core/services/ocr2/plugins/ocr2keeper/evm"
-	pluginevm "github.com/smartcontractkit/chainlink/v2/core/services/ocr2/plugins/ocr2keeper/evm"
```
Comment (infiloop2) [inline core/services/ocr2/plugins/ocr2keeper/evm20/log_provider.go:22]: just confirming @EasterTheBunny , plugin does not contain an evm package anymore?

File: core/services/ocr2/plugins/ocr2keeper/evm/abi.go [hunk 10]
```diff
@@ -3,17 +3,13 @@ package evm
...
 	"github.com/ethereum/go-ethereum/accounts/abi"
-	"github.com/ethereum/go-ethereum/common"
 	"github.com/ethereum/go-ethereum/common/hexutil"
 	ocr2keepers "github.com/smartcontractkit/ocr2keepers/pkg"
-)
 
-type evmRegistryPackerV2_0 struct {
-	abi abi.ABI
-}
+	iregistry21 "github.com/smartcontractkit/chainlink/v2/core/gethwrappers/generated/i_keeper_registry_master_wrapper_2_1"
```
[the remark this answers is not in the document]
Comment (amirylm) [reply core/services/ocr2/plugins/ocr2keeper/evm/abi.go:11]: i.e. to have 2.0 as `evm` and 2.1 as `evm21`? thought about it as well because there are less changes.

We said that because the final folder is called `evm` it be easier to work on the same files to avoid merging cross folders, but in general we can manage either way.

File: core/services/ocr2/plugins/ocr2keeper/evm20/log_provider.go (hunk 9 above)
Comment (amirylm) [reply core/services/ocr2/plugins/ocr2keeper/evm20/log_provider.go:22]: I moved log_provider into the evm folder, therefore the import is not needed

File: core/services/ocr2/plugins/ocr2keeper/evm/abi.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (infiloop2) [reply core/services/ocr2/plugins/ocr2keeper/evm/abi.go:11]: oh i was recommending evm20 and evm21. No confusion as to what evm stands for :sweat_smile:

File: core/services/ocr2/plugins/ocr2keeper/evm20/log_provider.go (hunk 9 above)
Comment (infiloop2) [reply core/services/ocr2/plugins/ocr2keeper/evm20/log_provider.go:22]: oh, looking at it more closely they were both referencing the exact same files :sweat_smile: . I thought there were separate 'evm' packages in core and plugin but seems like that's not the case anymore

Comment (EasterTheBunny) [reply core/services/ocr2/plugins/ocr2keeper/evm20/log_provider.go:22]: Right. I have tried to remove all evm stuff from the plugin. It should all be in the core node now.

**Review by EasterTheBunny — APPROVED** [review]

File: core/services/ocr2/plugins/ocr2keeper/evm/abi.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (amirylm) [reply core/services/ocr2/plugins/ocr2keeper/evm/abi.go:11]: Yeah, we can maybe go all the way and make it `evmregistry2x`

File: core/services/ocr2/plugins/ocr2keeper/evm/log_provider.go [hunk 11]
```diff
@@ -0,0 +1,483 @@
...
+			TransactionHash: p.TxHash.Hex(),
+			Confirmations:   end - p.BlockNumber,
+		}
+		vals = append(vals, l)
+	}
+
+	return vals, nil
+}
+
+func (c *LogProvider) StaleReportLogs(ctx context.Context) ([]ocr2keepers.StaleReportLog, error) {
```
Comment (RyanRHall) [inline core/services/ocr2/plugins/ocr2keeper/evm/log_provider.go:190]: nitpick, but I found this name confusing since it actually combines 3 log types and coerces them into one log type. Could we call this something generic like `AbortedReportLogs()` and make the return type generic? Not sure how annoying that would be. Not a strong opinion.

File: core/services/ocr2/plugins/ocr2keeper/evm/log_provider.go [hunk 12]
```diff
@@ -0,0 +1,483 @@
...
+			results = append(results, r)
+		}
+	}
+
+	return results, nil
+}
+
+// Fetches the checkBlockNumber for a particular transaction and an upkeep ID. Requires a RPC call to get txData
+// so this function should not be used heavily
+func (c *LogProvider) getCheckBlockNumberFromTxHash(txHash common.Hash, id ocr2keepers.UpkeepIdentifier) (bk ocr2keepers.BlockKey, e error) {
```
Comment (RyanRHall) [inline core/services/ocr2/plugins/ocr2keeper/evm/log_provider.go:417]: see comment above, but we should be able to ditch this with registry v2.1

File: core/services/ocr2/plugins/ocr2keeper/evm/registry.go [hunk 13]
```diff
@@ -65,12 +65,23 @@ var (
 
 //go:generate mockery --quiet --name Registry --output ./mocks/ --case=underscore
 type Registry interface {
-	GetUpkeep(opts *bind.CallOpts, id *big.Int) (keeper_registry_wrapper2_0.UpkeepInfo, error)
-	GetState(opts *bind.CallOpts) (keeper_registry_wrapper2_0.GetState, error)
+	GetUpkeep(opts *bind.CallOpts, id *big.Int) (iregistry21.UpkeepInfo, error)
+	GetState(opts *bind.CallOpts) (iregistry21.GetState, error)
 	GetActiveUpkeepIDs(opts *bind.CallOpts, startIndex *big.Int, maxCount *big.Int) ([]*big.Int, error)
+	GetActiveUpkeepIDsByType(opts *bind.CallOpts, startIndex *big.Int, endIndex *big.Int, trigger uint8) ([]*big.Int, error)
```
Comment (RyanRHall) [inline core/services/ocr2/plugins/ocr2keeper/evm/registry.go:71]: can we use the `upkeepType` enum here instead of `uint8` for trigger?

File: core/services/ocr2/plugins/ocr2keeper/evm20/abi.go [hunk 14]
```diff
@@ -0,0 +1,221 @@
...
+// enum UpkeepFailureReason
+// https://github.com/smartcontractkit/chainlink/blob/d9dee8ea6af26bc82463510cb8786b951fa98585/contracts/src/v0.8/interfaces/AutomationRegistryInterface2_0.sol#L94
+const (
+	UPKEEP_FAILURE_REASON_NONE = iota
+	UPKEEP_FAILURE_REASON_UPKEEP_CANCELLED
+	UPKEEP_FAILURE_REASON_UPKEEP_PAUSED
+	UPKEEP_FAILURE_REASON_TARGET_CHECK_REVERTED
+	UPKEEP_FAILURE_REASON_UPKEEP_NOT_NEEDED
+	UPKEEP_FAILURE_REASON_PERFORM_DATA_EXCEEDS_LIMIT
+	UPKEEP_FAILURE_REASON_INSUFFICIENT_BALANCE
```
Comment (RyanRHall) [inline core/services/ocr2/plugins/ocr2keeper/evm20/abi.go:27]: there's another one now, `CALLBACK_REVERTED`

File: core/services/ocr2/plugins/ocr2keeper/evm20/abi.go [hunk 15]
```diff
@@ -0,0 +1,221 @@
...
+		result.CheckBlockNumber = ret0.Result.CheckBlockNumber
+		result.CheckBlockHash = ret0.Result.CheckBlockhash
+		result.PerformData = ret0.Result.PerformData
+	}
+
+	// This is a default placeholder which is used since we do not get the execute gas
+	// from checkUpkeep result. This field is overwritten later from the execute gas
+	// we have for an upkeep in memory. TODO (AUTO-1482): Refactor this
+	result.ExecuteGas = 5_000_000
+
```
Comment (RyanRHall) [inline core/services/ocr2/plugins/ocr2keeper/evm20/abi.go:87]: I have a PR which returns the executeGas from `simulatePerformUpkeep` - do you need it returned from `checkUpkeep` also?

File: core/services/ocr2/plugins/ocr2keeper/evm20/abi.go [hunk 16]
```diff
@@ -0,0 +1,221 @@
...
+	if decoded, err = enc.DecodeReport(b); err != nil {
+		return nil, fmt.Errorf("error during decoding report while unpacking TransmitTxInput: %w", err)
+	}
+
+	return decoded, nil
+}
+
+var (
+	// rawPerformData is abi encoded tuple(uint32, bytes32, bytes). We create an ABI with dummy
+	// function which returns this tuple in order to decode the bytes
```
Comment (RyanRHall) [inline core/services/ocr2/plugins/ocr2keeper/evm20/abi.go:197]: not needed in v2.1

File: core/services/ocr2/plugins/ocr2keeper/evm20/encoder.go [hunk 17]
```diff
@@ -0,0 +1,230 @@
...
+	"github.com/smartcontractkit/ocr2keepers/pkg/encoding"
+)
+
+type EVMAutomationEncoder20 struct {
+	encoding.BasicEncoder
+}
+
+var (
+	Uint256, _                = abi.NewType("uint256", "", nil)
+	Uint256Arr, _             = abi.NewType("uint256[]", "", nil)
```
Comment (RyanRHall) [inline core/services/ocr2/plugins/ocr2keeper/evm20/encoder.go:19]: a helper like `mustNewType()` that panics on error might be nice :)

File: core/services/ocr2/plugins/ocr2keeper/evm20/encoder.go [hunk 18]
```diff
@@ -0,0 +1,230 @@
...
+	PerformDataMarshalingArgs = []abi.ArgumentMarshaling{
+		{Name: "checkBlockNumber", Type: "uint32"},
+		{Name: "checkBlockhash", Type: "bytes32"},
+		{Name: "performData", Type: "bytes"},
+	}
+	PerformDataArr, _   = abi.NewType("tuple(uint32,bytes32,bytes)[]", "", PerformDataMarshalingArgs)
```
Comment (RyanRHall) [inline core/services/ocr2/plugins/ocr2keeper/evm20/encoder.go:25]: see comment above, not needed in v2.1

File: core/services/ocr2/plugins/ocr2keeper/evm20/encoder.go [hunk 19]
```diff
@@ -0,0 +1,230 @@
...
+	ErrUnexpectedResult = fmt.Errorf("unexpected result struct")
+	packFn              = reportArgs.Pack
+	unpackIntoMapFn     = reportArgs.UnpackIntoMap
+	mKeys               = []string{"fastGasWei", "linkNative", "upkeepIds", "wrappedPerformDatas"}
+	reportArgs          = abi.Arguments{
+		{Name: mKeys[0], Type: Uint256},
+		{Name: mKeys[1], Type: Uint256},
+		{Name: mKeys[2], Type: Uint256Arr},
+		{Name: mKeys[3], Type: PerformDataArr},
```
Comment (RyanRHall) [inline core/services/ocr2/plugins/ocr2keeper/evm20/encoder.go:34]: this has changed, we added a triggers field and an execute gas field

File: core/services/ocr2/plugins/ocr2keeper/evm20/mercury_lookup.go [hunk 20]
```diff
@@ -0,0 +1,312 @@
...
+
+	"github.com/avast/retry-go/v4"
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/patrickmn/go-cache"
+
+	"github.com/smartcontractkit/chainlink/v2/core/gethwrappers/generated/keeper_registry_wrapper2_0"
```
Comment (RyanRHall) [inline core/services/ocr2/plugins/ocr2keeper/evm20/mercury_lookup.go:26]: why are we using 2_0 here? We should try to only use 2_1 with mercury lookup stuff

File: core/services/ocr2/plugins/ocr2keeper/evm20/registry.go [hunk 21]
```diff
@@ -0,0 +1,822 @@
...
+	// call will be made to the registry to obtain the most recent upkeep info and refresh this cache.
+	DefaultUpkeepExpiration = 10 * time.Minute
+	// DefaultCooldownExpiration decides how long a Mercury upkeep will be put in cool down for the first time. within
+	// 10 minutes, subsequent failures will result in double amount of cool down period.
+	DefaultCooldownExpiration = 5 * time.Second
+	// DefaultApiErrExpiration decides a running sum of total errors of an upkeep in this 10 minutes window. it is used
+	// to decide how long the cool down period will be.
+	DefaultApiErrExpiration = 10 * time.Minute
+	// CleanupInterval decides when the expired items in cache will be deleted.
+	CleanupInterval = 15 * time.Minute
```
Comment (RyanRHall) [inline core/services/ocr2/plugins/ocr2keeper/evm20/registry.go:47]: ideally we could control these by tweaking the offchainconfig set on the registry, so that in case we need to change something, we don't have to ship a new version of the node... how hard would that be? I don't think it's a must do, but good to understand if these values are a 1 or 2 way door.

Comment (amirylm) [reply core/services/ocr2/plugins/ocr2keeper/evm20/registry.go:47]: Sounds good, but this change is not related to 2.1 so let's do it separately

File: core/services/ocr2/plugins/ocr2keeper/evm20/mercury_lookup.go (hunk 20 above)
Comment (amirylm) [reply core/services/ocr2/plugins/ocr2keeper/evm20/mercury_lookup.go:26]: Thats the folder for 2.0 contract, you can.find mercury changes for 2.1 in #9528

File: core/services/ocr2/plugins/ocr2keeper/evm/abi.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (amirylm) [reply core/services/ocr2/plugins/ocr2keeper/evm/abi.go:175]: This function is used to decode an arbitrary triggerConfig of a log upkeep (identified by upkeepID).

Note that we anyway need to keep `getLogTriggerConfig` in the contract so the go wrappers structs will be generated.

File: core/services/ocr2/plugins/ocr2keeper/evm20/encoder.go (hunk 18 above)
Comment (amirylm) [reply core/services/ocr2/plugins/ocr2keeper/evm20/encoder.go:25]: Thats in 2.0 folder

File: core/services/ocr2/plugins/ocr2keeper/evm20/encoder.go (hunk 19 above)
Comment (amirylm) [reply core/services/ocr2/plugins/ocr2keeper/evm20/encoder.go:34]: Thats in 2.0 folder

File: core/services/ocr2/plugins/ocr2keeper/evm/abi.go [hunk 22]
```diff
@@ -51,10 +51,12 @@ func (rp *evmRegistryPackerV2_0) UnpackCheckResult(key ocr2keepers.UpkeepKey, ra
 		return result, err
 	}
 
-	result = EVMAutomationUpkeepResult20{
-		Block:    uint32(block.Uint64()),
-		ID:       id,
-		Eligible: true,
+	result = EVMAutomationUpkeepResult21{
+		Block:            uint32(block.Uint64()),
```
[the remark this answers is not in the document]
Comment (infiloop2) [reply core/services/ocr2/plugins/ocr2keeper/evm/abi.go:56]: Let's defer that to a separate decision when we finalise the contracts as this is linked to the gas overhead

File: core/services/ocr2/plugins/ocr2keeper/util.go [hunk 23]
```diff
@@ -57,11 +58,11 @@ func EVMProvider(db *sqlx.DB, chain evm.Chain, lggr logger.Logger, spec job.Job,
 	return keeperProvider, nil
 }
 
-func EVMDependencies(spec job.Job, db *sqlx.DB, lggr logger.Logger, set evm.ChainSet, pr pipeline.Runner, mc *models.MercuryCredentials) (evmrelay.OCR2KeeperProvider, *kevm.EvmRegistry, Encoder, *LogProvider, error) {
+func EVMDependencies(spec job.Job, db *sqlx.DB, lggr logger.Logger, set evm.ChainSet, pr pipeline.Runner, mc *models.MercuryCredentials) (evmrelay.OCR2KeeperProvider, *kevm20.EvmRegistry, Encoder, *kevm20.LogProvider, error) {
```
Comment (infiloop2) [inline core/services/ocr2/plugins/ocr2keeper/util.go:61]: nit. call it EVMDependencies20?

File: core/services/ocr2/plugins/ocr2keeper/util.go (hunk 2 above)
Comment (amirylm) [reply core/services/ocr2/plugins/ocr2keeper/util.go]: A ContractVersion field will be added to job spec once 2.1 is stable, with default value of 2.0 to support older job specs
